roscpp_azure_iothub/src/ros_azure_iothub_cpp_node.cpp

```diff
@@ -274,55 +282,167 @@ static void subscribeTopic(const char* topicName, ROS_Azure_IoT_Hub* iotHub)
     iotHub->subscribers.push_back( iotHub->nh.subscribe(topicName, 10, callback) );
 }
 
-//Current code does not keep track of topics already subscribed
+// Current code does not keep track of topics already subscribed
 static void deviceTwinCallback(DEVICE_TWIN_UPDATE_STATE update_state, const unsigned char* payLoad, size_t size, void* userContextCallback)
 {
     (void)update_state;
     (void)size;
-	    
+
     ROS_Azure_IoT_Hub* iotHub = (ROS_Azure_IoT_Hub*)userContextCallback;
 
     JSON_Value* root_value = json_parse_string((const char*)payLoad);
     JSON_Object* root_object = json_value_get_object(root_value);
-    JSON_Object* arrayObject = json_object_dotget_object (root_object, "desired.ros_relays");
+    JSON_Object* desired_object = json_object_dotget_object(root_object, "desired");
+    JSON_Object* ros_object = (desired_object != NULL) ? desired_object : root_object;
+
+    // Get topics for subscription
+    JSON_Object* arrayObject = json_object_dotget_object(ros_object, "ros_relays");
     size_t objectCount = json_object_get_count(arrayObject);
 
     for (size_t i = 0; i < objectCount; i++)
     {
-        printf("  %s \n", json_object_get_name(arrayObject, i));
+        ROS_INFO("  %s\r\n", json_object_get_name(arrayObject, i));
         JSON_Value* value = json_object_get_value_at(arrayObject, i);
         const char* topicToSubscribe = json_value_get_string(value);
-        printf("  %s \n", json_value_get_string(value));
+        ROS_INFO("  %s\r\n", json_value_get_string(value));
 
-        if (IsTopicAvailableForSubscribe(topicToSubscribe))
+        // Only subscribe the topic that is avaiable but not subscribed before
+        if (topicToSubscribe != NULL && IsTopicAvailableForSubscribe(topicToSubscribe) && std::find(iotHub->topicsToSubscribe.begin(), iotHub->topicsToSubscribe.end(), topicToSubscribe) == iotHub->topicsToSubscribe.end())
         {
+            ROS_INFO("Subscribe topic:  %s\r\n", topicToSubscribe);
             iotHub->topicsToSubscribe.push_back(topicToSubscribe);
+            subscribeTopic(topicToSubscribe, iotHub);
         }
     }
 
-    std::unique_lock<std::mutex> lck(iotHub->mtx);
-    iotHub->signal = 1;
-    iotHub->cv.notify_all();
+    if (!g_is_startup)
+    {
+        // Note: We ignore the existing ros_dynamic_configurations settings during ros_azure_iothub_cpp_node starts up.
+        // Reason: Suppose the other nodes are working with the expected dynamic configurations,
+        // and no need to update their configurations with the old values.
```

Comment:
I am little bit confused by this. Mind to elaborate more?

Reply:
Thanks for the comments. Code updated to remove the global variable to make the logic clear.